### PR TITLE
Add DELETE_DEFAULT_BRANCH access check

### LINKED ISSUE #3401: Prevent accidental deletion of the default branch
In production environments it must not be possible to delete the default branch.

This is probably an easy fix in terms of code, but might involve touching a bunch of tests.
For the tests, it should be fine to re-assign the default branch to "beginning of time" instead of deleting + re-creating it.

### PR DESCRIPTION
**Pull request by XN137**
closes #3401 

first commit was my initial approach.
after some Discussion with Robert we thought handling everything in the access checker might better.

preventing accidental deletion now has to be done by setting up the right rules though (its not prevented out of the box, when there are 0 rules) ?

### CODE REVIEW

File: servers/services/src/main/java/org/projectnessie/services/authz/Check.java [hunk 1]
```diff
@@ -42,6 +42,7 @@
     CREATE_REFERENCE(true, false),
     ASSIGN_REFERENCE_TO_HASH(true, false),
     DELETE_REFERENCE(true, false),
+    DELETE_DEFAULT_BRANCH(false, false),
```
Comment (XN137) [inline servers/services/src/main/java/org/projectnessie/services/authz/Check.java:45]: due to its relation to `DELETE_REFERENCE` i put this right after (and did the same in all other files).
if we want (or need due to relying on enumeration number somewhere) new values to be added at the end, i can adjust them accordingly.

File: design/authorization.md [hunk 2]
```diff
@@ -69,6 +69,7 @@ which looks like this:
       BatchAccessChecker canViewReference(NamedRef ref);
       BatchAccessChecker canCreateReference(NamedRef ref);
       BatchAccessChecker canDeleteReference(NamedRef ref);
+      BatchAccessChecker canDeleteDefaultBranch();
```
Comment (XN137) [inline design/authorization.md:72]: imo its a bit odd that we link to the code but also copy its contents here

File: servers/quarkus-server/src/main/resources/application.properties [hunk 3]
```diff
@@ -60,6 +60,8 @@ nessie.server.send-stacktrace-to-client=false
 #   && role=='admin_user'
 # nessie.server.authorization.rules.allow_listing_reflog=\
 #   op=='VIEW_REFLOG' && role=='admin_user'
+# nessie.server.authorization.rules.allow_deleting_default_branch=\
+#   op=='DELETE_DEFAULT_BRANCH' && role=='admin_user'
```
Comment (XN137) [inline servers/quarkus-server/src/main/resources/application.properties:64]: imo its a bit odd that `VIEW_REFLOG` is not part of `allow_all` rule... but i followed this "new pattern" here.

Comment (snazy) [reply servers/quarkus-server/src/main/resources/application.properties:64]: Generally, the CEL expressions and rules are flexible and users can create/use what they need/want.

The examples are probably a bit misleading.

The VIEW_REFLOG example should probably use a different role name (superuser? gc-user?) and the example for DELETE_DEFAULT_BRANCH maybe use the "superuser" role. IDK.
But honestly, the "test_user" role names above are similar.

File: servers/services/src/main/java/org/projectnessie/services/authz/Check.java (hunk 1 above)
Comment (snazy) [reply servers/services/src/main/java/org/projectnessie/services/authz/Check.java:45]: That's fine here - we neither store nor use this enum's ordinal.

File: servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImplWithAuthorization.java [hunk 4]
```diff
@@ -124,7 +124,12 @@ protected void assignReference(NamedRef ref, String oldHash, Reference assignTo)
   @Override
   protected void deleteReference(NamedRef ref, String hash)
       throws NessieConflictException, NessieNotFoundException {
-    startAccessCheck().canDeleteReference(ref).checkAndThrow();
+    BatchAccessChecker check = startAccessCheck();
+    if (getConfig().getDefaultBranch().equals(ref.getName())) {
+      check.canDeleteDefaultBranch();
+    }
+    check.canDeleteReference(ref);
```
Comment (XN137) [inline servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImplWithAuthorization.java:131]: note: we keep the old check in place as well

**Review by snazy — APPROVED** [review]
+1

**Review by nastra — APPROVED** [review]
LGTM, but this is only preventing the delete when authorization is enabled, right? A user could still just delete the default branch when no authorization is enabled

File: servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImplWithAuthorization.java (hunk 4 above)
Comment (ajantha-bhat) [reply servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImplWithAuthorization.java:131]: why do we need to check both here? This should be in else case ?

Comment (ajantha-bhat) [reply servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImplWithAuthorization.java:131]: Basically for default reference, if the user has access to delete it then it should not matter whether user has delete reference permission or not right?

@nastra, @snazy : WDYT ?

File: servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImplWithAuthorization.java [hunk 5]
```diff
@@ -124,7 +124,12 @@ protected void assignReference(NamedRef ref, String oldHash, Reference assignTo)
   @Override
   protected void deleteReference(NamedRef ref, String hash)
       throws NessieConflictException, NessieNotFoundException {
-    startAccessCheck().canDeleteReference(ref).checkAndThrow();
+    BatchAccessChecker check = startAccessCheck();
+    if (getConfig().getDefaultBranch().equals(ref.getName())) {
```
Comment (snazy) [inline servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImplWithAuthorization.java:128]: Actually, you need to check whether `ref` is a `BranchName`

Comment (XN137) [reply servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImplWithAuthorization.java]: addressed in latest commit

File: servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImplWithAuthorization.java (hunk 4 above)
Comment (XN137) [reply servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImplWithAuthorization.java]: moved to else after confirming with Robert

**Review by ajantha-bhat — APPROVED** [review]

**Comment by XN137** [comment]
> LGTM, but this is only preventing the delete when authorization is enabled, right? A user could still just delete the default branch when no authorization is enabled

yes indeed, which is why i mention this in the PR description:

> preventing accidental deletion now has to be done by setting up the right rules though (its not prevented out of the box, when there are 0 rules) ?

i am looking for ideas how we can now make it more difficult to delete the default branch out of the box :sweat_smile:

**Review by snazy — DISMISSED** [review]

**Comment by XN137** [comment]
until we have a good idea how to prevent default branch deletion out of the box i suggest to rename this PR to just `Add DELETE_DEFAULT_BRANCH access check`, wdyt?

**Comment by ajantha-bhat** [comment]
> until we have a good idea how to prevent default branch deletion out of the box i suggest to rename this PR to just Add DELETE_DEFAULT_BRANCH access check, wdyt?

Just thinking out loud, 
when the access checker is enabled, we can follow the current code.
when the access checker is disabled, we need to know whether the access checker was there or directly executing. we can use a thread local to set some flag when access checker is present. If the thread local flag is present. That means already validated and allow deleting. Else always block deleting inside  `TreeApiImpl#deleteReference`

**Comment by XN137** [comment]
PR is renamed and needs another review after recent test fixes

**Pull request merged**